pkg/storage/chunkfs/scanner.go

```diff
@@ -0,0 +1,362 @@
+	s.logger.Infof("starting watcher()")
+	defer s.logger.Infof("exiting from watcher()")
+
+	if s.cfg.SyncWorkers <= 0 {
+		s.logger.Infof("the number of SyncWorkers in the config is zero or negative, exiting from the watcher")
+		return
+	}
+
+	s.watcherWithF(ctx, s.scanDir)
+}
+
+func (s *Scanner) watcherWithF(ctx context.Context, f func(ctx context.Context, dir string)) {
+	for {
+		start := time.Now()
+		fis := files.ListDir(s.cfg.DataPath)
+		var wg sync.WaitGroup
+		sema := make(chan struct{}, s.cfg.SyncWorkers)
+		for _, fi := range fis {
+			// ignore all not-dirs
+			if !fi.IsDir() {
+				continue
+			}
+			sema <- struct{}{}
+			wg.Add(1)
+			go func(dir string) {
+				defer wg.Done()
+				defer func() {
+					<-sema
+				}()
+				f(ctx, dir)
+			}(filepath.Join(s.cfg.DataPath, fi.Name()))
+		}
+		wg.Wait()
+		diff := time.Now().Sub(start)
+		if diff < s.cfg.GlobalSyncTimeout {
+			select {
+			case <-ctx.Done():
+				// game over. Everything seemst to be stopped
+				return
+			case <-time.After(s.cfg.GlobalSyncTimeout - diff):
+				// ok, no leaks here, because it is the main case for the flow
+			}
+		}
+	}
+}
+
+func newScanInfos() scanInfos {
+	return make(scanInfos)
+}
+
+func (s *Scanner) scanDir(ctx context.Context, dir string) {
+	sis := s.buildScanInfos(dir)
+
+	sfn := filepath.Join(dir, cScanFileName)
+	if err := sis.readFromFile(sfn); err != nil {
+		s.logger.Warnf("could not read data from the file %s: %s", sfn, err)
+	}
+
+	s.sweep(ctx, dir, sis)
+	s.replicate(ctx, sis)
+	sis.writeToFile(sfn)
+}
+
+func (s *Scanner) buildScanInfos(dir string) scanInfos {
+	sis := newScanInfos()
+	for _, fi := range files.ListDir(dir) {
+		if fi.IsDir() {
+			continue
+		}
+		var si scanInfo
+		if doesLookLikeID(fi.Name()) {
+			si = scanInfo{ID: fi.Name(), Mtime: fi.ModTime(), size: fi.Size()}
+		} else if isItScanInfoFile(fi.Name()) {
+			var err error
+			sifn := filepath.Join(dir, fi.Name())
+			si, err = readScanInfo(sifn)
+			_ = os.Remove(sifn)
+			if err != nil {
+				s.logger.Warnf("could not read scanInfo from %s: %s", sifn, err)
+				continue
+			}
+		} else {
+			continue
+		}
+
+		// merge the syncInfo (si) with what we have so far about the chunk ID
+		if oldSI, ok := sis[si.ID]; ok {
+			if oldSI.Mtime.After(si.Mtime) {
+				si.Mtime = oldSI.Mtime
+			}
+			if si.SyncTime == nil {
+				si.SyncTime = oldSI.SyncTime
+			}
+			si.size = max(si.size, oldSI.size)
+		}
+		sis[si.ID] = si
+	}
+	return sis
+}
+
+func (s *Scanner) sweep(ctx context.Context, dir string, sis scanInfos) {
+	size := sis.size()
+	if size < s.cfg.SweepMaxThresholdSize {
+		// do nothing, we did not hit the local limit yet
+		return
+	}
+	csis := container.Values(sis)
+	s.logger.Infof("sweeping %s, current size=%d has %d chunks", dir, size, len(csis))
+	sort.Slice(csis, func(i, j int) bool {
+		return csis[i].Mtime.Before(csis[j].Mtime)
+	})
+	swept := 0
+	for _, si := range csis {
+		err := s.Replicator.DeleteChunk(ctx, si.ID, RFRemoteSync)
```

Comment:
thoughts: Looks a bit confusing to me that the replicator which should do the replication has the `DeleteChunk()` method which seems to be mainly responsible for deleting chunks (not replicating them) and only if you set some specific flag it will also replicate the chunk. Also, the current context is `sweeping` and not `replication`, which happens next, but it seems we're doing both here...

Reply:
DeleteChunk is a little bit overloaded: it deletes chunk on the local file system and may do one of two other things with the remote chunk:
1. Remove it from the remote storage as well
2. Upload/update the remote copy by the local one before removing it locally


It is made for to address the two scenarios:
1 remove the chunk from the local storage but keep it remotely
2 remove the chunk permanently from everywhere

Reply:
I understand that there was a reason or two, just mentioned what I think. Thanks for explaining.